# T81-558: Applications of Deep Neural Networks
**Module 2: Python for Machine Learning**
* Instructor: [Jeff Heaton](https://sites.wustl.edu/jeffheaton/), McKelvey School of Engineering, [Washington University in St. Louis](https://engineering.wustl.edu/Programs/Pages/default.aspx)
* For more information visit the [class website](https://sites.wustl.edu/jeffheaton/t81-558/).

# Module 2 Material

Main video lecture:

* Part 2.1: Introduction to Pandas [[Video]](https://www.youtube.com/watch?v=bN4UuCBdpZc&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN) [[Notebook]](t81_558_class_02_1_python_pandas.ipynb)
* Part 2.2: Categorical Values [[Video]](https://www.youtube.com/watch?v=4a1odDpG0Ho&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN) [[Notebook]](t81_558_class_02_2_pandas_cat.ipynb)
* Part 2.3: Grouping, Sorting, and Shuffling in Python Pandas [[Video]](https://www.youtube.com/watch?v=YS4wm5gD8DM&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN) [[Notebook]](t81_558_class_02_3_pandas_grouping.ipynb)
* **Part 2.4: Using Apply and Map in Pandas for Keras** [[Video]](https://www.youtube.com/watch?v=XNCEZ4WaPBY&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN) [[Notebook]](t81_558_class_02_4_pandas_functional.ipynb)
* Part 2.5: Feature Engineering in Pandas for Deep Learning in Keras [[Video]](https://www.youtube.com/watch?v=BWPTj4_Mi9E&list=PLjy4p-07OYzulelvJ5KVaT2pDlxivl_BN) [[Notebook]](t81_558_class_02_5_pandas_features.ipynb)

# Part 2.4: Apply and Map

The **apply** and **map** functions can also be applied to Pandas **dataframes**.

### Using Map with Dataframes

In [1]:
import os
import pandas as pd
import numpy as np

df = pd.read_csv(
    "https://data.heatonresearch.com/data/t81-558/auto-mpg.csv", 
    na_values=['NA', '?'])

display(df[0:10])

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0  18.0          8         307.0       130.0    3504          12.0    70   
1  15.0          8         350.0       165.0    3693          11.5    70   
2  18.0          8         318.0       150.0    3436          11.0    70   
3  16.0          8         304.0       150.0    3433          12.0    70   
4  17.0          8         302.0       140.0    3449          10.5    70   
5  15.0          8         429.0       198.0    4341          10.0    70   
6  14.0          8         454.0       220.0    4354           9.0    70   
7  14.0          8         440.0       215.0    4312           8.5    70   
8  14.0          8         455.0       225.0    4425          10.0    70   
9  15.0          8         390.0       190.0    3850           8.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino  
5       1           ford galaxie 500  
6       1           chevrolet impala  
7       1          plymouth fury iii  
8       1           pontiac catalina  
9       1         amc ambassador dpl

In [2]:
df['origin_name'] = df['origin'].map({1: 'North America', 2: 'Europe', 3: 'Asia'})
display(df[0:50])

mpg  cylinders  displacement  horsepower  weight  acceleration  year  \
0   18.0          8         307.0       130.0    3504          12.0    70   
1   15.0          8         350.0       165.0    3693          11.5    70   
2   18.0          8         318.0       150.0    3436          11.0    70   
3   16.0          8         304.0       150.0    3433          12.0    70   
4   17.0          8         302.0       140.0    3449          10.5    70   
5   15.0          8         429.0       198.0    4341          10.0    70   
6   14.0          8         454.0       220.0    4354           9.0    70   
7   14.0          8         440.0       215.0    4312           8.5    70   
8   14.0          8         455.0       225.0    4425          10.0    70   
9   15.0          8         390.0       190.0    3850           8.5    70   
10  15.0          8         383.0       170.0    3563          10.0    70   
11  14.0          8         340.0       160.0    3609           8.0    70   
12  15.0          8         400.0       150.0    3761           9.5    70   
13  14.0          8         455.0       225.0    3086          10.0    70   
14  24.0          4         113.0        95.0    2372          15.0    70   
15  22.0          6         198.0        95.0    2833          15.5    70   
16  18.0          6         199.0        97.0    2774          15.5    70   
17  21.0          6         200.0        85.0    2587          16.0    70   
18  27.0          4          97.0        88.0    2130          14.5    70   
19  26.0          4          97.0        46.0    1835          20.5    70   
20  25.0          4         110.0        87.0    2672          17.5    70   
21  24.0          4         107.0        90.0    2430          14.5    70   
22  25.0          4         104.0        95.0    2375          17.5    70   
23  26.0          4         121.0       113.0    2234          12.5    70   
24  21.0          6         199.0        90.0    2648          15.0    70   
25  10.0          8         360.0       215.0    4615          14.0    70   
26  10.0          8         307.0       200.0    4376          15.0    70   
27  11.0          8         318.0       210.0    4382          13.5    70   
28   9.0          8         304.0       193.0    4732          18.5    70   
29  27.0          4          97.0        88.0    2130          14.5    71   
30  28.0          4         140.0        90.0    2264          15.5    71   
31  25.0          4         113.0        95.0    2228          14.0    71   
32  25.0          4          98.0         NaN    2046          19.0    71   
33  19.0          6         232.0       100.0    2634          13.0    71   
34  16.0          6         225.0       105.0    3439          15.5    71   
35  17.0          6         250.0       100.0    3329          15.5    71   
36  19.0          6         250.0        88.0    3302          15.5    71   
37  18.0          6         232.0       100.0    3288          15.5    71   
38  14.0          8         350.0       165.0    4209          12.0    71   
39  14.0          8         400.0       175.0    4464          11.5    71   
40  14.0          8         351.0       153.0    4154          13.5    71   
41  14.0          8         318.0       150.0    4096          13.0    71   
42  12.0          8         383.0       180.0    4955          11.5    71   
43  13.0          8         400.0       170.0    4746          12.0    71   
44  13.0          8         400.0       175.0    5140          12.0    71   
45  18.0          6         258.0       110.0    2962          13.5    71   
46  22.0          4         140.0        72.0    2408          19.0    71   
47  19.0          6         250.0       100.0    3282          15.0    71   
48  18.0          6         250.0        88.0    3139          14.5    71   
49  23.0          4         122.0        86.0    2220          14.0    71   

    origin                          name    origin_name  
0        1     chev

### Using Apply with Dataframes

If the **apply** function is directly executed on the data frame, the lambda function is called once per column or row, depending on the value of axis.  For axis = 1, rows are used. 

The following code calculates a series called **efficiency** that is the **displacement** divided by **horsepower**. 

In [3]:
effi = df.apply(lambda x: x['displacement']/x['horsepower'], axis=1)
display(effi[0:10])

0    2.361538
1    2.121212
2    2.120000
3    2.026667
4    2.157143
5    2.166667
6    2.063636
7    2.046512
8    2.022222
9    2.052632
dtype: float64

### Feature Engineering with Apply and Map

In this section we will see how to calculate a complex feature using map, apply, and grouping.  The data set is the following CSV:

* https://www.irs.gov/pub/irs-soi/16zpallagi.csv 

This is US Government public data for "SOI Tax Stats - Individual Income Tax Statistics".  The primary website is here:

* https://www.irs.gov/statistics/soi-tax-stats-individual-income-tax-statistics-2016-zip-code-data-soi 

Documentation describing this data is at the above link.

For this feature, we will attempt to estimate the adjusted gross income (AGI) for each of the zipcodes.  The data file contains many columns; however, you will only use the following:

* STATE - The state (e.g. MO)
* zipcode - The zipcode (e.g. 63017)
* agi_stub - Six different brackets of annual income (1 through 6) 
* N1 - The number of tax returns for each of the agi_stubs

Note, the file will have 6 rows for each zipcode, for each of the agi_stub brackets. You can skip zipcodes with 0 or 99999.

We will create an output CSV with these columns; however, only one row per zip code. Calculate a weighted average of the income brackets. For example, the following 6 rows are present for 63017:


|zipcode |agi_stub | N1 |
|--|--|-- |
|63017	 |1 | 4710 |
|63017	 |2 | 2780 |
|63017	 |3 | 2130 |
|63017	 |4 | 2010 |
|63017	 |5 | 5240 |
|63017	 |6 | 3510 |


We must combine these six rows into one.  For privacy reasons, AGI's are broken out into 6 buckets.  We need to combine the buckets and estimate the actual AGI of a zipcode. To do this, consider the values for N1:

* 1 = \$1 to \$25,000
* 2 = \$25,000 to \$50,000
* 3 = \$50,000 to \$75,000
* 4 = \$75,000 to \$100,000
* 5 = \$100,000 to \$200,000
* 6 = \$200,000 or more

The median of each of these ranges is approximately:

* 1 = \$12,500
* 2 = \$37,500
* 3 = \$62,500 
* 4 = \$87,500
* 5 = \$112,500
* 6 = \$212,500

Using this you can estimate 63017's average AGI as:

```
>>> totalCount = 4710 + 2780 + 2130 + 2010 + 5240 + 3510
>>> totalAGI = 4710 * 12500 + 2780 * 37500 + 2130 * 62500 + 2010 * 87500 + 5240 * 112500 + 3510 * 212500
>>> print(totalAGI / totalCount)

88689.89205103042
```

In [4]:
import pandas as pd

df=pd.read_csv('https://www.irs.gov/pub/irs-soi/16zpallagi.csv')

First, we trim all zipcodes that are either 0 or 99999.  We also select the three fields that we need.

In [5]:
df=df.loc[(df['zipcode']!=0) & (df['zipcode']!=99999),['STATE','zipcode','agi_stub','N1']]

In [6]:
df

STATE  zipcode  agi_stub    N1
6         AL    35004         1  1510
7         AL    35004         2  1410
8         AL    35004         3   950
9         AL    35004         4   650
10        AL    35004         5   630
...      ...      ...       ...   ...
179785    WY    83414         2    40
179786    WY    83414         3    40
179787    WY    83414         4     0
179788    WY    83414         5    40
179789    WY    83414         6    30

[179184 rows x 4 columns]

We replace all of the **agi_stub** values with the correct median values with the **map** function.

In [7]:
medians = {1:12500,2:37500,3:62500,4:87500,5:112500,6:212500}
df['agi_stub']=df.agi_stub.map(medians)

In [8]:
df

STATE  zipcode  agi_stub    N1
6         AL    35004     12500  1510
7         AL    35004     37500  1410
8         AL    35004     62500   950
9         AL    35004     87500   650
10        AL    35004    112500   630
...      ...      ...       ...   ...
179785    WY    83414     37500    40
179786    WY    83414     62500    40
179787    WY    83414     87500     0
179788    WY    83414    112500    40
179789    WY    83414    212500    30

[179184 rows x 4 columns]

Next the dataframe is grouped by zip code.

In [9]:
groups = df.groupby(by='zipcode')

A lambda is applied across the groups and the AGI estimate is calculated.

In [10]:
df = pd.DataFrame(groups.apply(lambda x:sum(x['N1']*x['agi_stub'])/sum(x['N1']))).reset_index()

In [11]:
df

zipcode             0
0         1001  52895.322940
1         1002  64528.451001
2         1003  15441.176471
3         1005  54694.092827
4         1007  63654.353562
...        ...           ...
29867    99921  48042.168675
29868    99922  32954.545455
29869    99925  45639.534884
29870    99926  41136.363636
29871    99929  45911.214953

[29872 rows x 2 columns]

The new agi_estimate column is renamed.

In [12]:
df.columns = ['zipcode','agi_estimate']

In [13]:
display(df[0:10])

zipcode  agi_estimate
0     1001  52895.322940
1     1002  64528.451001
2     1003  15441.176471
3     1005  54694.092827
4     1007  63654.353562
5     1008  57575.757576
6     1009  45576.923077
7     1010  61303.191489
8     1011  49807.692308
9     1012  53214.285714

We can also see that our zipcode of 63017 gets the correct value.

In [14]:
df[ df['zipcode']==63017 ]

zipcode  agi_estimate
19909    63017  88689.892051